### DISCUSSION TITLE
sycl_ext_intel_device_info support?

### DISCUSSION THREAD

**Discussion by al42and** [AdaptiveCpp/AdaptiveCpp#1569] [Ideas]
Hey all!

I was wondering if there would be any interest in supporting [`sycl_ext_intel_device_info`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_device_info.md) extension in AdaptiveCpp. In particular, the vendor-agnostic queries like UUID and PCI Address.

--
Best,
Andrey

**Comment by illuhad** [comment]
Hi Andrey,

What is the intended use case? Do you want to uniquely identify devices within one backend (e.g. to be independent of things like `CUDA_VISIBLE_DEVICES` and similar), or would you want to to be able to identify devices across backends?

For the former, it is very likely that something like this will appear because we need it for some of the scheduling experiments we want to do.
For the latter, it's a bit more questionable because it really only makes sense for a subset of devices - devices may not be PCI devices, some OpenCL implementations may not support it etc.

Independently of this, it is highly unlikely that we will be implementing any oneAPI extension in an API-compatible way in the foreseeable future. Intel recently has set the precedent that when extensions from other implementations are added, this should be in an API-incompatible way even when there is no technical reason for this
(https://github.com/intel/llvm/blob/bfc4015a3f858ef664e469a9a0dbda97c2d22a1f/sycl/doc/extensions/experimental/sycl_ext_codeplay_enqueue_native_command.asciidoc).

I think this is silly, anti-user, and I've pushed against this, but it is what it is. As long as Intel does not change their stance here, I don't think we can endorse any oneAPI extensions in an API-compatible way.

**Reply by al42and** [reply]
> What is the intended use case? Do you want to uniquely identify devices within one backend (e.g. to be independent of things like CUDA_VISIBLE_DEVICES and similar), or would you want to to be able to identify devices across backends?

- The most pressing concern for us now is identifying devices within one backend across processes in the presence of `CUDA_VISIBLE_DEVICES` settings. E.g., SLURM/user sets `CUDA_VISIBLE_DEVICES` per MPI rank: we want each rank to use the designated device, but we also want to know whether it's the same device used by all ranks or not. See https://gitlab.com/gromacs/gromacs/-/issues/5065.
  - One can argue whether `CUDA_VISIBLE_DEVICES` is the right method to use. But that's what [JUWELS-Booster does automatically](https://apps.fz-juelich.de/jsc/hps/juwels/gpu-computing.html#gpu-visibility-affinity) and what [LUMI recommends](https://docs.lumi-supercomputer.eu/runjobs/scheduled-jobs/lumig-job/). Also, for out-of-order execution, there's still #1321.
- Related to that, optimal CPU-GPU mapping can be [complex](https://docs.lumi-supercomputer.eu/runjobs/scheduled-jobs/distribution-binding/#gpu-binding), and logging which device (PCI address) is used by which rank will help ensuring the correct affinity.
- Being able to tell if it's the same device available through two backends is also nice. Yes, it won't necessary work in all cases, but when it works it can avoid the confusion.
- Having device UUID in the log can help identifying glitchy devices.
- In addition to the device identification, "free global memory" also could be of broad interest.

> For the former, it is very likely that something like this will appear because we need it for some of the scheduling experiments we want to do.

Nice! Any hope of it being in 24.10?

> I think this is silly, anti-user, and I've pushed against this, but it is what it is. As long as Intel does not change their stance here, I don't think we can endorse any oneAPI extensions in an API-compatible way.

I am not sure that being as anti-user as your competitor is a good strategy :) But if that's how it is, ok, we can definitely handle a different API.

**Reply by illuhad** [reply]
Okay, thanks for the feedback :)

> Nice! Any hope of it being in 24.10?

I dunno. I had too many vacation days accumulated, so I was gone for most of september and large portions of August, and I've only returned today. So there's still a lot of other feature work on the table for 24.10 that needs to be done :(

> I am not sure that being as anti-user as your competitor is a good strategy :) But if that's how it is, ok, we can definitely handle a different API.

I agree, I hate it. But I feel like this is the only leverage that we have in AdaptiveCpp to put pressure on Intel for this issue... The end goal here is not to be anti-user, but to hope for a change in DPC++.

**Reply by al42and** [reply]
> I dunno. I had too many vacation days accumulated, so I was gone for most of september and large portions of August, and I've only returned today. So there's still a lot of other feature work on the table for 24.10 that needs to be done :(

Hope you had nice vacation!

Any objection to the proposal below? Can push an implementation up next week.

```markdown
### `ACPP_EXT_DEVICE_NATIVE_UUID`

This extension allows querying the UUID as returned by the backend. It adds `info::device::ext_AdaptiveCpp_native_uuid`
information descriptor with return type `std::array<std::byte, 16>`. If the device does not support querying the UUID,
the output will contain all zeros. Please note that even when the returned value is non-zero, it is not guaranteed to
conform to any of the standards for UUID or even be _universally unique_ the vendor library might return an identical
hardcoded value for all devices.
```

We can also add an aspect to check the availability or throw if the device does not support, but the user will still have to carefully evaluate their assumptions about UUIDs, and returning all-zeroes is in-line with the current industry practices:
- AMD consumer(?) devices always have the same (non-zero) hardcoded value,
- Intel uses a combination of {0x8086, device PCI ID, device PCI revision, PCI address} (possibly with messed-up endianness), which is not globally unique,
- AMD datacenter(?) devices have non-standard-compliant value (they convert 8-byte binary ID to 16 hex-digits in ASCII, and return those 16 bytes, which leads to inconsistent [_variant_](https://en.wikipedia.org/wiki/Universally_unique_identifier#Variants) value), that is at least globally unique.

**Reply by illuhad** [reply]
>Hope you had nice vacation!

Thank you :)

> Any objection to the proposal below? Can push an implementation up next week.

Thanks for looking into this! Some feedback:

- `info::device::ext_AdaptiveCpp_native_uuid` - I think we don't prefix extension APIs with `ext` in code, so `AdaptiveCpp_native_uuid` would be sufficient. Perhaps to avoid confusion with actual UUIDs (which the definition later says this is not), better call it `AdaptiveCpp_native_id` or `AdaptiveCpp_unique_id`?
- Do we really need 128 bit? PCI device ID AFAIK is 32-bit, so would `uint64_t` not be sufficient to store a combination of device ID/bus ID/ backend ID? `uint64_t` might be easier to handle than `array<byte, 16>` for users.
- `it is not guaranteed to conform to any of the standards for UUID or even be _universally unique_ the vendor library might return an identical hardcoded value for all devices.` - I'm not sure I understand this limitation. Does this not mean that we basically cannot rely on any guarantees from this query? Why can we not say that it needs to obey guarantees XYZ, otherwise it will be 0?
- I think there may be some different use cases here, which might benefit from slightly different variants of this query. E.g.
  - Request unique device ID when you want to identify physical devices across different backends. This can be useful for workload distribution.
  - Request unique ID that distinguishes different backends, where the same device would return a different ID if it is from different backends. This will be useful e.g. for performance profiling since different backends may have different performance characteristics.

So perhaps we need two different queries?

**Reply by al42and** [reply]
> I think there may be some different use cases here, which might benefit from slightly different variants of this query.
> ...
> So perhaps we need two different queries?

Even three.

> Request unique device ID when you want to identify physical devices across different backends. This can be useful for workload distribution.

That would be very hard to do properly. OpenMP devices on different MPI ranks on the same / different nodes? Several POSIX threads, each with its own OpenMP device, pinned to different sets of cores? (If I recall correctly, mixing POSIX threads and OpenMP in one process is technically UB, but we do it and it has worked fine so far). Two processes, each with one RDNA GPU (with no vendor UUID support) with the identical PCI Address -- same device or two different devices on two nodes?

I don't see a practical way to do it with any reasonable guarantees. With PCIe Address, we can guarantee that two SYCL devices that both have a valid ID and are on the same node are separate physical devices; perhaps we can mix in somehow the node identifier? But that still leaves the possibility that a device with such an ID can be the same as another device with no ID (if there's no way to query the address).

> Request unique ID that distinguishes different backends, where the same device would return a different ID if it is from different backends. This will be useful e.g. for performance profiling since different backends may have different performance characteristics.

As far as I can tell, that requires solving the first problem, and then mixing in the backend ID. 

What I want to achieve here is different. A way to match the physical device across different backends, _including the ability to compare with the native vendor API_. Using the same "UUID" as reported by the driver does that decently:
- For NVIDIA GPUs: CUDA, nvidia-smi/NVML, and OpenCL report consistent values;
- For AMD GPUs: HIP and rocm-smi report consistent values; no UUID support in AMD OpenCL;
- For Intel GPUs: LevelZero and OpenCL (both Intel's and Rusticl) report consistent values; `xpu-smi` reports a similar value with a different byte order (00000000-0000-0003-0000-000856a08086 vs 8680a056-0800-0000-0300-000000000000, :facepalm:).

Another way to achieve this would be using PCI Addresses, with as similar set of limitations. The difference is that the consumer AMD GPUs would work, but NVIDIA IDs won't be globally-unique anymore. And it will fit into 64 bits.

<hr/>


> I think we don't prefix extension APIs with `ext` in code, so `AdaptiveCpp_native_uuid` would be sufficient.

No strong preference from me, but [the spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_names_for_extensions_to_existing_classes_or_enumerations) suggests using it:

> Vendors are encouraged to use the `ext_<vendorstring>_` prefix form when possible for additions to existing SYCL classes because this form makes the extension’s vendor name apparent..

> Perhaps to avoid confusion with actual UUIDs (which the definition later says this is not), better call it AdaptiveCpp_native_id or AdaptiveCpp_unique_id?

Calling it `unique_id` is equally misleading, at least until we find a way to ensure it's actually unique. `native_id` is easy to confuse with the ordinal id. At least "UUID" is how AMD/Intel/NVIDIA call the things in their API; `native` shifts the blame onto the vendor :)

> I'm not sure I understand this limitation. Does this not mean that we basically cannot rely on any guarantees from this query? Why can we not say that it needs to obey guarantees XYZ, otherwise it will be 0?

We can say that it matches the "UUID" reported by the device driver (if any).

Without vendor support, promising global uniqueness is a big challenge. Promising local (within the node) uniqueness is easier, but might require departing from what the vendor API reports, so the value will only be usable within ACpp.

**Reply by illuhad** [reply]
> I don't see a practical way to do it with any reasonable guarantees. With PCIe Address, we can guarantee that two SYCL devices that both have a valid ID and are on the same node are separate physical devices; perhaps we can mix in somehow the node identifier? But that still leaves the possibility that a device with such an ID can be the same as another device with no ID (if there's no way to query the address).

Right, I did not consider the multi-node use case. I would be fine with only guaranteeing uniqueness within the node. SYCL is a solution for intra-node parallelism, if you use multiple nodes I think it is fair to expect the user to handle this additional complexity.

I suspect though that the multi-node case could be handled by returning a larger id, e.g. prepending additional 64-bits for the node hash or so. But as I said, for me this would not be a critical feature for now.

OpenMP might be a candidate for just returning 0/no id for this query. The only overlap can be with OpenCL implementations, and there we might have the problem that the mapping to cores might be different as you say. E.g. OpenCL might expose CPUs in multi-socket systems as separate devices.

> As far as I can tell, that requires solving the first problem, and then mixing in the backend ID.

Yes, but perhaps with some nuance. For example, I would say that in this case we could still assign an Id to the OpenMP device, since we can neatly distinguish it from other CPU device abstractions, like an OpenCL CPU device.

> What I want to achieve here is different. A way to match the physical device across different backends, including the ability to compare with the native vendor API. Using the same "UUID" as reported by the driver does that decently:

Okay, so you want more of an interop id? But how would it be guaranteed that you can actually match physical devices if we at the same time also need to return the values directly reported by the driver? Different backend drivers may return different things.

> Calling it unique_id is equally misleading, at least until we find a way to ensure it's actually unique. native_id is easy to confuse with the ordinal id. At least "UUID" is how AMD/Intel/NVIDIA call the things in their API; native shifts the blame onto the vendor :)

Yeah, we might have to do some bikeshedding, especially for all the different variants of the query.

FWIW, AdaptiveCpp `sycl::device` already has `AdaptiveCpp_device_id` which returns the internal runtime device identifier:

https://github.com/AdaptiveCpp/AdaptiveCpp/blob/8cf7a90232d065b80813eedfae678518fa67eed7/include/hipSYCL/sycl/device.hpp#L260

I think for the purpose of distinguishing multiple device software abstractions across backends there could be an `AdaptiveCpp_stable_device_id`, which is the same as `AdaptiveCpp_device_id`, except that it is stable w.r.t environment variables like `CUDA_VISIBLE_DEVICES`.

Then there could be something like `physical_id` which refers to the physical device, `local_unique_id`(or uuid), `unique_physical_id` or similar. Or `native_(uu)id` for the interop thing.

> Without vendor support, promising global uniqueness is a big challenge. Promising local (within the node) uniqueness is easier, but might require departing from what the vendor API reports, so the value will only be usable within ACpp.

Yes, I agree. I'm primarily concerned with use cases within the AdaptiveCpp infrastructure. For your use case, perhaps a different approach is needed; perhaps with different return types to match native vendor APIs.

> No strong preference from me, but [the spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_names_for_extensions_to_existing_classes_or_enumerations) suggests using it:

Yes, but this is only to ensure that SYCL WG does not introduce some future feature that then has name collisions with extensions. The actual naming recommendation was changed multiple times and it's possible that the current one is not the final one either.
I think that `AdaptiveCpp_*` as prefix already serves the purpose of preventing name collisions, so the `ext_*` is superfluous. And we also don't use it in our current interfaces.

**Reply by al42and** [reply]
> SYCL is a solution for intra-node parallelism, if you use multiple nodes I think it is fair to expect the user to handle this additional complexity.

True, but that depends on what we promise. If we say "unique ID", my understanding would be that it is globally unique. Locally-unique ID would be just "ID"; if it's not locally unique, how is it even an "identifier" at all?

> Okay, so you want more of an interop id?

Yes, although I don't like the name. "interop id", to me, implies the "device ordinal" (e.g., the one passed to `cudaSetDevice`).

> But how would it be guaranteed that you can actually match physical devices if we at the same time also need to return the values directly reported by the driver? Different backend drivers may return different things.

Yes, but if we look through interop lens, we can only interoperate with a single driver. If two different drivers can be used for the same device and they are not interoperable with each other and don't offer a way to consistently identify the device, we cannot help them.

But, so far, the `native_uuid` approach appears to do it for NVIDIA and Intel across whatever backends I had installed (PoCL for CUDA is the only relevant one missing that I can think of). AMD is the only major GPU vendor where it does not work at all; but then, their OpenCL implementation won't allow querying PCIe address either.

>  I think that AdaptiveCpp_* as prefix already serves the purpose of preventing name collisions, so the ext_* is superfluous. And we also don't use it in our current interfaces.

Ok, makes sense.

<hr/>

So, trying to summarize, there are several dimensions here:

- Physical vs. logical device
- Locally vs globally unique
- Interoperable or ACpp-specific
- For logical devices: stable vs unstable

And the options discussed are:

- "True" unique ID: some unique global ID for the physical device. ACpp-specific, hard to implement.
- Native "UUID": UUID reported by the backend driver for the physical device. Interoperable with backend, easy to implement but possibly missing or non-unique even locally.
- PCIe address: Physical device, locally unique, interoperable with backend, easy to implement but possibly missing.
- AdaptiveCpp_device_id: ACpp-specific, implemented, logical device, non-stable, locally-unique, always available.
- AdaptiveCpp_stable_device_id: ACpp-specific, logical device, stable, locally-unique, always available.
- AdaptiveCpp_physical_device_id: ACpp-specific, physical device, locally-unique, always available.
- We can generate ACpp-specific, logical device, non-stable, globally-unique, always available identifiers, but those are not very useful?
- Don't think we can have stable, globally-unique IDs for logical devices.

Am I missing something above?

**Comment by blinkfrog** [comment]
I would also find this useful. In my experience, there are cases where application needs to associate a `sycl::device` with information, obtained through platform-specific APIs (such as DXGI on Windows). For this, a stable physical/native device identifier would be very valuable.
